src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -158,15 +212,14 @@ class RaftIvfIndexNode : public IndexNode {
                     return metric.error();
                 }
                 if (metric.value() != raft::distance::DistanceType::L2Expanded &&
-                    metric.value() != raft::distance::DistanceType::L2Unexpanded &&
                     metric.value() != raft::distance::DistanceType::InnerProduct) {
                     LOG_KNOWHERE_WARNING_ << "selected metric not supported in RAFT IVF indexes: "
                                           << ivf_raft_cfg.metric_type;
                     return Status::invalid_metric_type;
                 }
-
+                devs_.insert(devs_.begin(), ivf_raft_cfg.gpu_ids.begin(), ivf_raft_cfg.gpu_ids.end());
                 auto scoped_device = detail::device_setter{*ivf_raft_cfg.gpu_ids.begin()};
-                res_ = std::make_unique<raft::device_resources>();
+                auto res_ = raft_res_pool::resource::instance().get_raft_res(devs_[0]);
```

Comment:
I notice a stream-pool being created above. Is the goal to parallelize multiple queries and saturate the GPU by using different cuda streams or different threads? What guarantee does `Knowhere` make wrt the number of threads/gpus which might be invoking `search()` concurrently to compute the QPS?

If Knowhere is using different threads (and GPUs) to perform queries then we absolutely should be able to submit the queries on different streams to saturate each GPU. Though- I think we will want to be careful to make sure each GPU has its own copy of the IVF indexes in that case. 

We are going to need to set the stream from the stream pool explicitly on a copy of the `device_resources` instance for each stream in the stream pool, though. As it works right now, the `device_resources` can contain its own stream pool, which allows algorithms to use it internally when they support it. It's not automatic, though. If you are using `default_stream_per_thread` then I would create a new `device_resources` instance for each thread (maybe change `context()` to `thread_local` in that case?

Reply:
knowhere have a thread pool, now search concurrent size is the cpu core number.

Reply:
@Presburger okay so it sounds like multiple threads might be using the `raft::device_resources` resources but are all of the 
 the instances being created by the same thread?